Updated node design with versioning

krynble opened the pull request.

ahsanv committed:
:zap: introduce versioned nodes

ahsanv committed:
Export versioned nodes for separate process run

ahsanv committed:
Add bse node for versioned nodes

ahsanv committed:
fix node name for versioned nodes

ahsanv committed:
extend node from nodeVersionedType

ahsanv committed:
improve nodes base and flow to FE

ahsanv committed:
revert lib es2019 to es2017

ahsanv committed:
include version in key to prevent duplicate key

ahsanv committed:
handle type versions on FE

ahsanv committed:
clean up

ahsanv committed:
cleanup nodes base

ahsanv committed:
add type versions in getNodeParameterOptions

ahsanv committed:
cleanup

ahsanv committed:
code review

ahsanv committed:
code review + add default version to node type description

ahsanv committed:
remove node default types from store

ahsanv committed:
:lipstick: cleanups

krynble committed:
Merge master

krynble committed:
Draft for migrated Mattermost node

krynble committed:
First version of Mattermost node versioned according to node standards

krynble committed:
Correcting deactivate operations name to match currently used one

krynble committed:
Merge branch 'master' into versioned-nodes-updated

ivov commented on packages/workflow/src/Interfaces.ts:
```diff
@@ -482,6 +486,10 @@ export interface ITriggerResponse {
 	manualTriggerResponse?: Promise<INodeExecutionData[][]>;
 }
 
+export interface INodeVersions {
+	[key: number]: INodeType;
+}
+
```
`INodeVersionedType` is not using this at L514. Is it not the same?

ivov commented on packages/workflow/src/Interfaces.ts:
```diff
@@ -502,6 +510,15 @@ export interface INodeType {
+	nodeVersions: {
+		[key: number]: INodeType
+	};
+	defaultVersion: number;
```
Ideally only positive integers. Same for the arg in `getNodeType`. I will look into whether TS can enforce this.

ivov replied:
Not possible at the moment.

ivov commented on packages/nodes-base/src/NodeVersionedType.ts:
```diff
@@ -0,0 +1,21 @@
+
+	constructor(nodeVersions: INodeVersions, description: INodeTypeBaseDescription, defaultVersion?: number) {
+		this.nodeVersions = nodeVersions;
+		this.defaultVersion = defaultVersion || Math.max(...Object.keys(nodeVersions).map(k => parseInt(k, 10)));
```
1. Ideally `??` instead of `||`.
2. Another option: `.map(Number)`
3. The keys of `nodesVersions` are already numbers. Why convert them into numbers?
4. Since `defaultVersion` can be the default version or the latest version, maybe `currentVersion` would be a more fitting key name.

ivov commented on packages/nodes-base/nodes/Mattermost/Mattermost.node.ts:
```diff
@@ -1,2396 +1,26 @@
-
-				return returnData;
-			},
+		nodeVersions: INodeVersions = {
+			1: new MattermostV1(description),
```
If we have many versions, are instantiating them all?

ivov commented on packages/nodes-base/nodes/Mattermost/Mattermost.node.ts:
```diff
@@ -1,2396 +1,26 @@
+			group: ['output'],
+			subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
+			description: 'Sends data to Mattermost',
+			defaultVersion: 1,
```
Why do we need to specify the default version...? 
1. in the `description` at L18, 
2. in the assignment at L23, and 
3. as the third arg to `super()` at L24

I would refactor to have it only in the `description` and use it always from there - but maybe I am missing something.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/MattermostV1.node.ts:
```diff
@@ -0,0 +1,30 @@
+	constructor(baseDescription: INodeTypeBaseDescription) {
+		this.description = {
+			...baseDescription,
+			...description,
```
Another option: `nodeDescription` and `nodeProperties`

ivov commented on packages/nodes-base/nodes/Mattermost/v1/transport/index.ts:
```diff
@@ -22,7 +22,7 @@ export interface IAttachment {
 
 /**
- * Make an API request to Telegram
+ * Make an API request to Mattermost
```
This is a good opportunity to standardize the removal of JSDoc arg descriptions.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/transport/index.ts:
```diff
@@ -34,7 +34,7 @@ export async function apiRequest(this: IHookFunctions | IExecuteFunctions | ILoa
 	const credentials = this.getCredentials('mattermostApi');
 
 	if (credentials === undefined) {
-		throw new NodeOperationError(this.getNode(), 'No credentials got returned!');
+		throw new NodeOperationError(this.getNode(), 'No credentials returned!');
```
We could refactor `this.getCredentials` to throw this error inside, so that we do not have to repeat this check in each and every node that has credentials.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/methods/loadOptions.ts:
```diff
@@ -0,0 +1,150 @@
+	apiRequest,
+} from '../transport';
+
+const loadOptions = {
```
We can do away with the wrapping object and its methods, have just functions at top level, and from `index.ts` we `export * from './loadOptions.ts'`. It would be simpler and in keeping with other files that only contain standalone functions.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/router.ts:
```diff
@@ -0,0 +1,53 @@
+				operationResult.push(...await reaction[mattermost.operation].call(this, i));
+			} else if (mattermost.resource === 'user') {
+				operationResult.push(...await user[mattermost.operation].call(this, i));
+			} else {
+				throw new NodeApiError(this.getNode(), {message: 'Resource not supported.'});
```
1. `NodeOperationError` instead of `NodeApiError`. 
2. In fact, at this `else` clause we have exhausted the union (if you try to access `mattermost.resource`, its type will be `never`) so instead of catching the error, we can rely on the IDE/TS to disallow us from using a non-existing resource.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/router.ts:
```diff
@@ -0,0 +1,53 @@
+	const operationResult: INodeExecutionData[] = [];
+
+	for (let i = 0; i < items.length; i++) {
+		const resource = this.getNodeParameter('resource', i) as string;
+		let operation = this.getNodeParameter('operation', i) as string;
```
The return types from `getNodeParameter` can and probably should be narrowed down. If we add to `getNodeParameter` two overloads for `resource` and `operation`, which are always a `string`, then we can remove the two castings here and convert the casting on L29 to a typing, to improve type safety.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/router.ts:
```diff
@@ -0,0 +1,53 @@
+		}
+	}
+
+	return [operationResult];
```
Should we array-ify the result here or back at `execute()`? Just a thought.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/router.ts:
```diff
@@ -0,0 +1,53 @@
+
+export async function router(this: IExecuteFunctions): Promise<INodeExecutionData[][]> {
+	const items = this.getInputData();
+	const operationResult: INodeExecutionData[] = [];
```
Subjective, but `operationResult` is quite wordy for a variable to be used so often. The original was `returnData`, which has `data` in the name, so probably best avoided as a var name. Maybe simply `output`? Thinking of future contributors and newcomers to the codebase.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/router.ts:
```diff
@@ -0,0 +1,53 @@
+			operation = 'deactive';
+		}
+		const mattermost = {
+			resource,
+			operation,
```
We can destructure here to further simplify below.

ivov commented on packages/nodes-base/nodes/Mattermost/v1/actions/user/create.ts:
```diff
@@ -0,0 +1,50 @@
+	const returnData: IDataObject[] = [];
+	const responseData = await apiRequest.call(this, requestMethod, endpoint, body, qs);
+
+	if (Array.isArray(responseData)) {
+		returnData.push(...responseData);
+	} else {
+		returnData.push(responseData);
+	}
+
+	return this.helpers.returnJsonArray(returnData);
```
Before, we had this pattern once per node - now it will multiply by the number of node actions. Therefore, can we tuck away this logic inside one of the helpers? Same comment applies to array-ifying the output at the end of every router.

ivov committed:
:sparkles: Create utility types

ivov committed:
:zap: Simplify Mattermost types

ivov committed:
:zap: Rename exports for consistency

ivov committed:
:zap: Type channel properties

ivov committed:
:zap: Type message properties

ivov committed:
:zap: Type reaction properties

ivov committed:
:zap: Type user properties

ivov committed:
:zap: Add type import to router

ivov committed:
:bug: Add missing key

ivov committed:
:hammer: Adjust typo in operation name

ivov committed:
:hammer: Inline exports for channel properties

ivov committed:
:hammer: Inline exports for message properties

ivov committed:
:hammer: Inline exports for reaction properties

ivov committed:
:hammer: Inline exports for user properties

ivov committed:
:hammer: Inline exports for load options

ivov committed:
:shirt: Fix lint issue

ivov committed:
:hammer: Inline export for description

ivov committed:
:hammer: Rename descriptions for clarity

ivov committed:
:hammer: Refactor imports/exports for methods

ivov committed:
:hammer: Refactor latest version retrieval

ivov committed:
:fire: Remove unneeded else clause

When the string literal union is exhausted, the resource key becomes never, so TS disallows wrong key usage.

ivov committed:
:sparkles: Add overloads to getNodeParameter

ivov committed:
:zap: Improve overload

ivov committed:
:fire: Remove superfluous INodeVersions type

ivov committed:
:hammer: Relocate pre-existing interface

ivov committed:
:fire: Remove JSDoc arg descriptions

ivov committed:
:zap: Minor reformatting in transport file

ivov committed:
:zap: Fix API call function type

krynble committed:
Merge pull request #1963 from n8n-io/ivov/nodes-redesign-contributions

Nodes redesign contributions

krynble committed:
Created first draft for Axios requests

krynble committed:
Merge branch 'versioned-nodes-updated' of github.com:n8n-io/n8n into versioned-nodes-updated

krynble committed:
Working version of mattermost node with Axios

ivov commented on packages/nodes-base/nodes/Mattermost/v1/transport/index.ts:
```diff
@@ -19,9 +21,9 @@ import {
 export async function apiRequest(
 	this: IHookFunctions | IExecuteFunctions | ILoadOptionsFunctions,
-	method: string,
+	method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'HEAD',
```
And `PATCH`

krynble committed:
Work in progress for replacing request library

krynble committed:
Merging master

krynble committed:
Improvements to request translations

krynble committed:
Fixed sending files via multipart / form-data

krynble committed:
Merging master

krynble committed:
Fixing translation from request to axios and loading node parameter options

krynble committed:
Merging master

krynble committed:
Improved typing for new http helper

krynble committed:
Added ignore any for specific lines for linting

krynble committed:
Fixed follow redirects changes on http request node and manual execution of previously existing workflow with older node versions

krynble committed:
Adding default headers according to body on httpRequest helper

krynble committed:
Spec error handling and fixed workflows with older node versions

krynble committed:
Showcase how to export errors in a standard format

krynble committed:
Merge branch 'master' into versioned-nodes-updated

krynble committed:
Merging master

krynble committed:
Merging master

krynble committed:
Merge branch 'master' into versioned-nodes-updated

krynble committed:
Refactored mattermost node to keep files in a uniform structure. Also fix bugs with merges

krynble committed:
Reverting changes to http request node

krynble committed:
Changed nullish comparison and removed repeated code from nodes

krynble committed:
Renamed queryString back to qs and simplified node output

krynble committed:
Simplified some comparisons

krynble committed:
Changed header names to be uc first

krynble committed:
Added default user agent to requests and patch http method support

krynble committed:
Fixed indentation, remove unnecessary file and console log

krynble committed:
Updating with master

krynble committed:
Merging master

janober commented on packages/nodes-base/nodes/Mattermost/v1/actions/channel/del/execute.ts:
```diff
@@ -0,0 +1,25 @@
+	apiRequest,
+} from '../../../transport';
+
+export async function del(this: IExecuteFunctions, index: number): Promise<INodeExecutionData[]> {
```
Why is this function and the folder called "del" and not "delete"?

krynble replied:
We named it `del` because `delete` is used for the [delete operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete) and was causing build problems.

janober commented on packages/nodes-base/nodes/Mattermost/Mattermost.node.ts:
```diff
@@ -1,2404 +1,27 @@
+export class Mattermost extends NodeVersionedType {
+	constructor() {
+		const baseDescription: INodeTypeBaseDescription = {
+			displayName: 'Mattermost V1',
```
Was this on purpose left with "V1"?

krynble replied:
Good catch, fixed the name.

krynble committed:
Merge branch 'master' into versioned-nodes-updated

krynble committed:
Fixed mattermost node name

krynble committed:
Fixed lint issues

krynble committed:
Further fix linting issues

krynble committed:
Further fix lint issues

krynble committed:
Fixed http request helper's return type

janober commented:
Thanks a lot. Got merged.

janober commented:
Got released with n8n@0.139.0
